File: 79.Word Search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        m = len(board)
        n = len(board[0])
        vis = [[False for j in range(n)] for i in range(m)]
        def dfs(i, j, k):
            if i <0 or i >=m or j<0 or j>=n or vis[i][j] or board[i][j]!=word[k]:
                return False
            k+=1
            if k == len(word):
                return True
            vis[i][j] = True
            if dfs(i-1, j, k) or dfs(i+1, j, k) or dfs(i, j-1, k) or dfs(i, j+1, k):
                return True
            vis[i][j] = False
            return False
        
        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        
        return False
        
        

File: 79. Word Search.go
func exist(board [][]byte, word string) bool {
	m, n := len(board), len(board[0])
	visited := make([]int, m*n)

	var existHelper func(i, j, k int) bool
	existHelper = func(i, j, k int) bool {
		if i < 0 || i >= m || j < 0 || j >= n || visited[i*n+j] == 1 || board[i][j] != word[k] {
			return false
		}

		if k == len(word)-1 {
			return true
		}

		visited[i*n+j] = 1
		if existHelper(i-1, j, k+1) || existHelper(i+1, j, k+1) || existHelper(i, j-1, k+1) || existHelper(i, j+1, k+1) {
			return true
		} else {
			visited[i*n+j] = 0
			return false
		}
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == word[0] && existHelper(i, j, 0) {
				return true
			}
		}
	}
	return false
}
